### ISSUE TITLE
Pointer math bug/limits

### ISSUE THREAD

**Issue by davepl** [cc65/cc65#2060]
This code works perfectly until n becomes more than about  32768, which makes me suspect a pointer math bug.  As long as n is only 16 bits in value, it works fine.  Any thoughts, or is it a compiler bug?

void SETBIT(byte *p, unsigned long n)
{
   *(p + (n >> 3)) |=  (0b10000000 >> (n & 7));
}

**Comment by mrdudz** [comment]
First, please give a full minimal example to demonstrate the problem, and tell exactly what happens, and what you expect to happen.

Also, are you sure this isnt the result of integer promotion?

This seems to be correct at least:
```
#include <stdio.h>

unsigned char *test(unsigned char *p, unsigned long n)
{
    return p + (n >> 3);
}

int main(void)
{
    printf("%p\n", test(0x1234, 0x12340000UL));
    printf("%p\n", test(0x1234, 0xba980000UL));
}

$ cl65 -t sim6502 -o bug2060.prg bug2060.c
$ sim65 bug2060.prg
0x9234
0x1234
```

**Comment by davepl** [comment]
My bad for not including the link to the code example.  Here it is:

https://github.com/davepl/cc65/blob/KIMLife/samples/kim1/kimGFX.c￼
cc65/kimGFX.c at KIMLife · davepl/cc65
github.com

The line draw which calls SETPIXEL can’t draw more than 4K into the screen buffer.  Yet ClearScreen() works fine.

To work around it I rewrote it as follows, which works.  But it seems to me both should work, unless I’ve simply got a but in the GitHub version above.  But given the arbitrary nature of the cutoff, it seemed odd.

- Dave

￼

**Comment by mrdudz** [comment]
Sorry, but this is far too complex to debug anything - hence "minimal example". Ideally strip it down to a single expression that does not work as intended.

**Comment by kugelfuhr** [comment]
The problem is not in the snippet shown. It is in the surrounding code and caused by an error in the code generation of the compiler

The problematic code is

```C
void SETPIXEL(byte * p, int x, int y, byte b)
{
   if (b)
      SETBIT(p, (long) y * SCREEN_WIDTH + x);
   else
      CLRBIT(p, (long) y * SCREEN_WIDTH + x);
}
```

This is correctly translated by an older cc65 version using "-T -Oirs" as:

```asm
;
; SETBIT(p, (long) y * SCREEN_WIDTH + x);
;
        ldy     #$08
        jsr     pushwysp                ; push p
        ldy     #$04
        lda     (sp),y
        tax
        dey
        lda     (sp),y                  ; load y
        jsr     axlong                  ; convert to long
        jsr     pusheax                 ; push (long) y
        ldx     #$01
        lda     #$00                    ; OOPS!
        lda     #$40                    ; SCREEN_WIDTH -> a/x
        jsr     tosmul0ax
        jsr     pusheax                 ; push (long) y * SCREEN_WIDTH
        ldy     #$0A
        lda     (sp),y
        tax
        dey
        lda     (sp),y                  ; load x
        jsr     axlong                  ; convert to long
        jsr     tosaddeax               ; calc (long) y * SCREEN_WIDTH + x
        jsr     _SETBIT
```

(comments by me).

The current cc65 version from git incorrectly translates this as follows:

```asm
;
; SETBIT(p, (long) y * SCREEN_WIDTH + x);
;
        ldy     #$08
        jsr     pushwysp                ; push p
        ldy     #$04
        lda     (sp),y
        tax
        dey
        lda     (sp),y                  ; load y
        jsr     axlong                  ; convert to long
        jsr     pusheax                 ; push (long) y
        ldx     #$01
        lda     #$00                    ; OOPS!
        lda     #$40                    ; SCREEN_WIDTH -> a/x
        jsr     tosmul0ax
        jsr     pusheax                 ; push (long) y * SCREEN_WIDTH
        jsr     axlong                  ; <= OOPSIE?!?! convert something to long
        ldy     #$0A
        lda     (sp),y
        tax
        dey
        lda     (sp),y                  ; load x                           
        jsr     tosaddeax               ; calculation has invalid hi 16 bit
        jsr     _SETBIT
```

As you can see, the generated code moves the long conversion of x before the actual load of x, which causes the high 16 bits of the value passed to SETBIT to be wrong. The error does not depend on the optimizer. Code compiled without -O has the same problem.

Please note the unnecessary load of A when loading SCREEN_WIDTH. This is also a regression. It has been introduced earlier than 2.18.

**Comment by davepl** [comment]
That’s OK, I’ll just fix it in my local fork, and hopefully someone more compliant with the way you like to work will run into it in the future, and you can fix this compiler regression in the main tree at some later date.

Cheers!
Dave

**Comment by mrdudz** [comment]
Well, if you are able to fix it - please make a PR. Chances are that noone else will.

**Comment by bbbradsmith** [comment]
Here is a minimal test that produces the bug:
[axlong_bug.zip](https://github.com/cc65/cc65/files/11375780/axlong_bug.zip)

Code:
```
// Test of bug: https://github.com/cc65/cc65/issues/2060

#include <stdio.h>

#define W 320

unsigned long test1(unsigned char* p, unsigned long n)
{
	return n;
}

unsigned long test0(unsigned char* p, int x, int y, unsigned char b)
{
	return test1(p, (long)y * W + x);
}

#define TEST(ta,tb) \
	expect = (long)tb * W + ta; \
	result = test0(p,ta,tb,0x56); \
	printf("%3d * %3d + %3d = %08lx",tb,W,ta,result); \
	if (expect != result) { printf(" expected: %08lx\n",expect); ++fail; } \
	else printf("\n");

int main(void)
{
	unsigned char* p = 0x1234;
	unsigned long expect, result;
	int fail = 0;
	
	TEST(1,3);
	TEST(50,60);
	TEST(99,88);
	TEST(128,102);
	TEST(129,102);
	TEST(320,102);
	TEST(0,103); // breaks when y > 102 ((y * 320) > 32767)
	TEST(150,170); // breaks
	TEST(300,180); // breaks
	return fail;
}
```

Output:
```
  3 * 320 +   1 = 000003c1
 60 * 320 +  50 = 00004b32
 88 * 320 +  99 = 00006e63
102 * 320 + 128 = 00008000
102 * 320 + 129 = 00008001
102 * 320 + 320 = 000080c0
103 * 320 +   0 = ffff80c0 expected: 000080c0
170 * 320 + 150 = ffffd516 expected: 0000d516
180 * 320 + 300 = ffffe22c expected: 0000e22c
```

**Comment by bbbradsmith** [comment]
Without optimization, this is what test0 looks like. It has the error identified by kugelfuhr above:
```
; return test1(p, (long)y * W + x);
.proc	_test0: near
	jsr     pusha
	ldy     #$06
	jsr     ldaxysp
	jsr     pushax
	ldy     #$04
	jsr     ldaxysp
	jsr     axlong
	jsr     pusheax
	ldx     #$01
	lda     #$00
	sta     sreg
	lda     #$00
	sta     sreg+1
	lda     #$40
	jsr     tosmuleax
	jsr     pusheax
	jsr     axlong
	ldy     #$0A
	jsr     ldaxysp
	jsr     tosaddeax
	jsr     _test1
	jmp     L0001
```

Things look fine through tosmuleax, pusheax. The 32-bit multiply result is on the stack.

At this point AX contains the low 16-bits of the multiply, and there is an "axlong" here which sign-extends this result into the high word. Then "x" is loaded, but X is not sign extended. Finally, "x" with this incorrect high word is added to the 32-bit result on the stack.

So, the axlong should have been after ldaxysp (fetch of int "x"). How does it accidentally end up before it?

**Comment by bbbradsmith** [comment]
I believe the problem is right here: [expr.c:3093](https://github.com/cc65/cc65/blob/805e98a7aa6fbd2b419de5192163f7d47c1f3beb/src/cc65/expr.c#L3093)
```
} else if (!DoArrayRef && IsClassInt (lhst) && IsClassInt (rhst)) {
	/* Integer addition */
	flags = typeadjust (Expr, &Expr2, 0);
	/* Load rhs into the primary */
	LoadExpr (CF_NONE, &Expr2);
} else {
```

The [comment for typeadjust](https://github.com/cc65/cc65/blob/805e98a7aa6fbd2b419de5192163f7d47c1f3beb/src/cc65/expr.c#L240):
```
/* Adjust the two values for a binary operation. lhs is expected on stack or
** to be constant, rhs is expected to be in the primary register or constant.
** The function will put the type of the result into lhs and return the
** code generator flags for the operation.
** If NoPush is given, it is assumed that the operation does not expect the lhs
** to be on stack, and that lhs is in a register instead.
** Beware: The function does only accept int types.
*/
```

I believe these two lines are simply out of order. typeadjust expects rhs to already be in EAX, but because LoadExpr hasn't yet happened, it's not in EAX.

Swapping them does seem to fix this example. Currently running unit tests.

**Comment by davepl** [comment]
Nice work!

**Comment by mrdudz** [comment]
fixed via #2070
